edb/server/protocol/ui_ext.pyx

```diff
@@ -56,32 +57,49 @@ async def handle_request(
     path_parts,
     server,
 ):
-    if path_parts == ['instance-info']:
-        # endpoint for data that either cannot be fetched by an edgeql query
-        # or is needed to make a connection to send queries
-        response.status = http.HTTPStatus.OK
-        response.content_type = b'application/json'
-        response.body = json.dumps({
-            'instance_name': server._instance_name if
-                server._instance_name is not None else
-                ('_localdev' if server.in_dev_mode() else None),
-            'databases': [
-                {'name': db.name}
-                for db in server._dbindex.iter_dbs()
-            ]
-        }).encode()
-        return
-
-    if path_parts == []:
-        urlpath = request.url.path.decode('ascii')
-        if urlpath.endswith('/'):
-            response.status = http.HTTPStatus.PERMANENT_REDIRECT
-            response.custom_headers['Location'] = urlpath[0:-1]
+    try:
+        if path_parts == ['instance-info']:
+            # endpoint for data that either cannot be fetched by an edgeql query
+            # or is needed to make a connection to send queries
+            response.status = http.HTTPStatus.OK
+            response.content_type = b'application/json'
+            response.body = json.dumps({
+                'instance_name': server._instance_name if
+                    server._instance_name is not None else
+                    ('_localdev' if server.in_dev_mode() else None),
+                'databases': [
+                    {'name': db.name}
+                    for db in server._dbindex.iter_dbs()
+                ],
+                'roles': list(server._roles.keys()),
+            }).encode()
             return
-        path_parts = ['index.html']
 
-    try:
-        data, content_type = static_files[os.path.join(*path_parts)]
+        if path_parts == []:
+            path_parts = ['index.html']
+
+            if not request.url.query:
+                ekey = server.get_jwe_key()
+                skey = server.get_jws_key()
+                claims = {"edgedb.server.any_role": True}
+                signed_token = jwt.JWT(header={"alg": "ES256"}, claims=claims)
+                signed_token.make_signed_token(skey)
+                token = jwt.JWT(
+                    header={"alg": "ECDH-ES", "enc": "A256GCM"},
+                    claims=signed_token.serialize(),
+                )
+                token.make_encrypted_token(ekey)
+                urlpath = request.url.path.decode('ascii')
+                response.status = http.HTTPStatus.TEMPORARY_REDIRECT
+                response.custom_headers['Location'] = (
+                    f'{urlpath}?authToken={token.serialize()}'
+                )
```

Comment:
Instead of doing the redirect here, can we send the token back in the `/ui/instance-info` request above?

Reply:
This redirect largely makes the token ineffective, as then nothing would stop a malicious script from fishing out the token and proceeding with making requests to the local instance.  Instead, the CLI must insert the token into the URL it generates in `edgedb ui`.

Reply:
Ah ok